java/ql/lib/semmle/code/java/os/OSCheck.qll

```diff
@@ -0,0 +1,93 @@
+/**
+ * Provides classes and predicates for guards that check for the current OS.
+ */
+
+import java
+import semmle.code.java.controlflow.Guards
+private import semmle.code.java.frameworks.apache.Lang
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.StringCheck
+
+/**
+ * A guard that checks if the current platform is Windows.
+ */
+abstract class IsWindowsGuard extends Guard { }
+
+/**
+ * A guard that checks if the current platform is unix or unix-like.
+ */
+abstract class IsUnixGuard extends Guard { }
+
+/**
+ * Holds when the MethodAccess is a call to check the current OS using either the upper case `osUpperCase` or lower case `osLowerCase` string constants.
+ */
+bindingset[osUpperCase, osLowerCase]
+private predicate isOsFromSystemProp(MethodAccess ma, string osUpperCase, string osLowerCase) {
+  exists(MethodAccessSystemGetProperty sgpMa |
+    sgpMa.hasCompileTimeConstantGetPropertyName("os.name")
+  |
+    DataFlow::localExprFlow(sgpMa, ma.getQualifier()) and // Call from System.getProperty to some partial match method
+    ma.getAnArgument().(CompileTimeConstantExpr).getStringValue().matches(osUpperCase)
+    or
+    exists(MethodAccess toLowerCaseMa |
+      toLowerCaseMa.getMethod() =
+        any(Method m | m.getDeclaringType() instanceof TypeString and m.hasName("toLowerCase"))
+    |
+      DataFlow::localExprFlow(sgpMa, toLowerCaseMa.getQualifier()) and // Call from System.getProperty to toLowerCase
+      DataFlow::localExprFlow(toLowerCaseMa, ma.getQualifier()) and // Call from toLowerCase to some partial match method
+      ma.getAnArgument().(CompileTimeConstantExpr).getStringValue().matches(osLowerCase)
+    )
+  ) and
+  isStringPartialMatch(ma)
+}
+
+private class IsWindowsFromSystemProp extends IsWindowsGuard instanceof MethodAccess {
+  IsWindowsFromSystemProp() { isOsFromSystemProp(this, "Window%", "window%") }
+}
+
+private class IsUnixFromSystemProp extends IsUnixGuard instanceof MethodAccess {
+  IsUnixFromSystemProp() {
+    isOsFromSystemProp(this, ["Mac%", "Linux%", "LINUX%"], ["mac%", "linux%"])
+  }
+}
+
+private predicate isOsFromApacheCommons(FieldAccess fa, string fieldName) {
+  exists(Field f | f = fa.getField() |
+    f.getDeclaringType() instanceof ApacheSystemUtils and
+    f.hasName(fieldName)
+  )
+}
+
+private class IsWindowsFromApacheCommons extends IsWindowsGuard instanceof FieldAccess {
+  IsWindowsFromApacheCommons() { isOsFromApacheCommons(this, "IS_OS_WINDOWS%") }
+}
+
+private class IsUnixFromApacheCommons extends IsUnixGuard instanceof FieldAccess {
+  IsUnixFromApacheCommons() { isOsFromApacheCommons(this, "IS_OS_UNIX") }
```

Comment:
Should this also consider Linux (`IS_OS_LINUX`) and Mac (`IS_OS_MAC%`) (and possibly some of the other Unix-like constants)?

Reply:
I've been thinking about this over the past few days, and I don't know if it can be in all cases. I'm trying to think through the logic. Take this example:

```java
    /**
     * Should only be called on windows
     */
    private void onlyOnWindows() {}
    
    /**
     * Should only be called on unix like systems
     */
    private void onlyOnUnix() {}


    void test() {
        if (System.getProperty("os.name").contains("Windows")) {
            onlyOnWindows();
        }

        if (SystemUtils.IS_OS_UNIX) {
            onlyOnUnix();
        } else {
            onlyOnWindows();
        }

        if (!SystemUtils.IS_OS_LINUX) {
            // Might be another different unix-like system, so this can't be `onlyOnWindows()`.
        } else {
            onlyOnUnix();
        }
    }
```

I think the same issue exists with `"IS_OS_WINDOWS%"` above too.  If `SystemUtils.IS_OS_WINDOWS_2000` evaluates to `false`, it doesn't mean that you're on a unix system.

I need to fix this more generally. Trying to wrap my head around how to do so. @smowton, any suggestions you have here would be appreciated.

Reply:
Sounds right to me: all the more specific tests are meaningful in the context of a positive finding but a negative one is uninformative. I would recommend doing this by categorising the various flags into those which indicates is-any-unix (which is roughly the negation of is-any-windows), vs. those that indicate is-specific-unix (and similarly is-specific-windows). It's then up to the configuration incorporating these into an isBarrierGuard predicate to only use is-specific-unix positively, while is-any-unix can be used in either direction.